File: src/gui/dialogs/export_dialog.py
"""
Enhanced Export Dialog for NC-KTV
Supports multiple export modes: Karaoke, Lyrics Video, etc.
"""

from PyQt6.QtWidgets import (
    QDialog, QVBoxLayout, QHBoxLayout, QLabel, 
    QComboBox, QPushButton, QCheckBox, QFileDialog,
    QLineEdit, QFormLayout, QGroupBox, QRadioButton,
    QButtonGroup, QColorDialog, QMessageBox
)
from PyQt6.QtGui import QColor
from PyQt6.QtCore import Qt
from pathlib import Path


class ExportDialog(QDialog):
    """Enhanced dialog for video export with multiple modes"""
    
    # Export mode presets
    MODE_KARAOKE_VIDEO = "karaoke_video"
    MODE_LYRICS_VIDEO = "lyrics_video"
    MODE_KARAOKE_NO_VIDEO = "karaoke_no_video"
    MODE_LYRICS_NO_VIDEO = "lyrics_no_video"
    MODE_CUSTOM = "custom"
    
    def __init__(self, parent=None, default_filename="karaoke.mp4", current_animation="Linear Wipe", project=None):
        super().__init__(parent)
        self.setWindowTitle("Export Video - NC-KTV")
        self.resize(650, 700)
        self.output_path = None
        self.current_animation = current_animation
        self.project = project
        self.background_color = QColor(20, 20, 30)  # Dark blue default
        
        self._init_ui(default_filename)
        
        # Set default mode
        self._apply_preset(self.MODE_KARAOKE_VIDEO)
    
    def _init_ui(self, default_filename):
        """Initialize UI"""
        layout = QVBoxLayout(self)
        
        # Title
        title = QLabel("🎬 Export Your Karaoke Video")
        title.setStyleSheet("font-size: 16pt; font-weight: bold; padding: 10px;")
        layout.addWidget(title)
        
        # === Preset Modes ===
        preset_group = QGroupBox("Quick Presets")
        preset_layout = QVBoxLayout()
        
        # Preset buttons
        btn_layout = QHBoxLayout()
        
        self.btn_karaoke_video = QPushButton("🎤 Karaoke Video\n(Music Video + Instrumental)")
        self.btn_karaoke_video.setCheckable(True)
        self.btn_karaoke_video.setMinimumHeight(60)
        self.btn_karaoke_video.clicked.connect(lambda: self._apply_preset(self.MODE_KARAOKE_VIDEO))
        btn_layout.addWidget(self.btn_karaoke_video)
        
        self.btn_lyrics_video = QPushButton("🎵 Lyrics Music Video\n(Music Video + Original Audio)")
        self.btn_lyrics_video.setCheckable(True)
        self.btn_lyrics_video.setMinimumHeight(60)
        self.btn_lyrics_video.clicked.connect(lambda: self._apply_preset(self.MODE_LYRICS_VIDEO))
        btn_layout.addWidget(self.btn_lyrics_video)
        
        preset_layout.addLayout(btn_layout)
        
        btn_layout2 = QHBoxLayout()
        
        self.btn_karaoke_no_video = QPushButton("📺 Karaoke (No Video)\n(Solid BG + Instrumental)")
        self.btn_karaoke_no_video.setCheckable(True)
        self.btn_karaoke_no_video.setMinimumHeight(60)
        self.btn_karaoke_no_video.clicked.connect(lambda: self._apply_preset(self.MODE_KARAOKE_NO_VIDEO))
        btn_layout2.addWidget(self.btn_karaoke_no_video)
        
        self.btn_lyrics_no_video = QPushButton("📝 Lyrics Video\n(Solid BG + Original Audio)")
        self.btn_lyrics_no_video.setCheckable(True)
        self.btn_lyrics_no_video.setMinimumHeight(60)
        self.btn_lyrics_no_video.clicked.connect(lambda: self._apply_preset(self.MODE_LYRICS_NO_VIDEO))
        btn_layout2.addWidget(self.btn_lyrics_no_video)
        
        preset_layout.addLayout(btn_layout2)
        
        # Button group for mutual exclusivity
        self.preset_group = QButtonGroup()
        self.preset_group.addButton(self.btn_karaoke_video, 0)
        self.preset_group.addButton(self.btn_lyrics_video, 1)
        self.preset_group.addButton(self.btn_karaoke_no_video, 2)
        self.preset_group.addButton(self.btn_lyrics_no_video, 3)
        
        preset_group.setLayout(preset_layout)
        layout.addWidget(preset_group)
        
        # === Custom Options ===
        options_group = QGroupBox("Custom Options")
        options_layout = QFormLayout()
        
        # Video Source
        self.combo_video_source = QComboBox()
        self.combo_video_source.addItems([
            "Music Video (if available)",
            "Solid Color Background",
            "Custom Image",
            "No Background (Transparent)"
        ])
        self.combo_video_source.currentIndexChanged.connect(self._on_video_source_changed)
        options_layout.addRow("Video Source:", self.combo_video_source)
        
        # Background Color (for solid bg)
        bg_layout = QHBoxLayout()
        self.btn_bg_color = QPushButton("Choose Color...")
        self.btn_bg_color.clicked.connect(self._choose_bg_color)
        self.lbl_bg_preview = QLabel("   ")
        self.lbl_bg_preview.setFixedSize(50, 25)
        self.lbl_bg_preview.setStyleSheet(f"background-color: {self.background_color.name()}; border: 1px solid #888;")
        bg_layout.addWidget(self.btn_bg_color)
        bg_layout.addWidget(self.lbl_bg_preview)
        bg_layout.addStretch()
        options_layout.addRow("Background Color:", bg_layout)
        
        # Audio Source
        self.combo_audio_source = QComboBox()
        self.combo_audio_source.addItems([
            "Instrumental Only (Karaoke)",
            "Original Audio (With Vocals)",
            "Instrumental + Vocals Mixed"
        ])
        options_layout.addRow("Audio Track:", self.combo_audio_source)
        
        # Animation Style
        animation_layout = QHBoxLayout()
        self.lbl_animation = QLabel(f"{current_animation}")
        self.lbl_animation.setStyleSheet("font-weight: bold; color: #4CAF50;")
        animation_layout.addWidget(self.lbl_animation)
        animation_layout.addWidget(QLabel("(from preview)"))
        animation_layout.addStretch()
        options_layout.addRow("Animation:", animation_layout)
        
        # Visual Style
        self.combo_style = QComboBox()
        self.combo_style.addItems(["Neon Gold", "Classic Blue", "Clean White", "Custom"])
        options_layout.addRow("Lyrics Style:", self.combo_style)
        
        options_group.setLayout(options_layout)
        layout.addWidget(options_group)
        
        # === Output File ===
        output_group = QGroupBox("Output File")
        output_layout = QFormLayout()
        
        file_layout = QHBoxLayout()
        self.txt_path = QLineEdit(default_filename)
        self.btn_browse = QPushButton("Browse...")
        self.btn_browse.clicked.connect(self._browse)
        file_layout.addWidget(self.txt_path)
        file_layout.addWidget(self.btn_browse)
        output_layout.addRow("Save As:", file_layout)
        
        output_group.setLayout(output_layout)
        layout.addWidget(output_group)
        
        layout.addStretch()
        
        # === Buttons ===
        btn_layout = QHBoxLayout()
        
        btn_cancel = QPushButton("Cancel")
        btn_cancel.clicked.connect(self.reject)
        btn_layout.addWidget(btn_cancel)
        
        btn_layout.addStretch()
        
        btn_export = QPushButton("🎬 Export Video")
        btn_export.clicked.connect(self._validate_and_accept)
        btn_export.setStyleSheet(
            "background-color: #E91E63; color: white; font-weight: bold; "
            "padding: 10px 20px; font-size: 12pt;"
        )
        btn_layout.addWidget(btn_export)
        
        layout.addLayout(btn_layout)
    
    def _apply_preset(self, mode):
        """Apply a preset mode"""
        self.current_mode = mode
        
        if mode == self.MODE_KARAOKE_VIDEO:
            # Music video BG + Instrumental audio
            self.combo_video_source.setCurrentIndex(0)  # Music Video
            self.combo_audio_source.setCurrentIndex(0)  # Instrumental
            self.btn_karaoke_video.setChecked(True)
            self.btn_karaoke_video.setStyleSheet("background-color: #4CAF50; color: white;")
            
        elif mode == self.MODE_LYRICS_VIDEO:
            # Music video BG + Original audio
            self.combo_video_source.setCurrentIndex(0)  # Music Video
            self.combo_audio_source.setCurrentIndex(1)  # Original
            self.btn_lyrics_video.setChecked(True)
            self.btn_lyrics_video.setStyleSheet("background-color: #4CAF50; color: white;")
            
        elif mode == self.MODE_KARAOKE_NO_VIDEO:
            # Solid BG + Instrumental
            self.combo_video_source.setCurrentIndex(1)  # Solid Color
            self.combo_audio_source.setCurrentIndex(0)  # Instrumental
            self.btn_karaoke_no_video.setChecked(True)
            self.btn_karaoke_no_video.setStyleSheet("background-color: #4CAF50; color: white;")
            
        elif mode == self.MODE_LYRICS_NO_VIDEO:
            # Solid BG + Original audio
            self.combo_video_source.setCurrentIndex(1)  # Solid Color
            self.combo_audio_source.setCurrentIndex(1)  # Original
            self.btn_lyrics_no_video.setChecked(True)
            self.btn_lyrics_no_video.setStyleSheet("background-color: #4CAF50; color: white;")
        
        # Reset other buttons
        for btn in [self.btn_karaoke_video, self.btn_lyrics_video, 
                    self.btn_karaoke_no_video, self.btn_lyrics_no_video]:
            if not btn.isChecked():
                btn.setStyleSheet("")
        
        self._on_video_source_changed()
    
    def _on_video_source_changed(self):
        """Enable/disable background color based on video source"""
        is_solid_bg = self.combo_video_source.currentIndex() == 1
        self.btn_bg_color.setEnabled(is_solid_bg)
        self.lbl_bg_preview.setVisible(is_solid_bg)
    
    def _choose_bg_color(self):
        """Choose background color"""
        color = QColorDialog.getColor(self.background_color, self, "Choose Background Color")
        if color.isValid():
            self.background_color = color
            self.lbl_bg_preview.setStyleSheet(
                f"background-color: {color.name()}; border: 1px solid #888;"
            )
    
    def _browse(self):
        """Browse for output file"""
        path, _ = QFileDialog.getSaveFileName(
            self, "Save Video", self.txt_path.text(), "MP4 Video (*.mp4)"
        )
        if path:
            self.txt_path.setText(path)
    
    def _validate_and_accept(self):
        """Validate and accept"""
        if not self.txt_path.text():
            QMessageBox.warning(self, "No Output File", "Please specify an output file.")
            return
        
        # Check if video source is available
        video_idx = self.combo_video_source.currentIndex()
        if video_idx == 0:  # Music video
            if not self.project or not self.project.source_file:
                QMessageBox.warning(
                    self,
                    "No Music Video",
                    "Music video source not available. Please choose a different video source."
                )
                return
        
        self.output_path = Path(self.txt_path.text())
        self.accept()
    
    def get_options(self):
        """Get export options"""
        video_idx = self.combo_video_source.currentIndex()
        audio_idx = self.combo_audio_source.currentIndex()
        
        return {
            "path": self.output_path,
            "mode": self.current_mode,
            "video_source": {
                0: "music_video",
                1: "solid_color",
                2: "custom_image",
                3: "transparent"
            }[video_idx],
            "audio_source": {
                0: "instrumental",
                1: "original",
                2: "mixed"
            }[audio_idx],
            "background_color": self.background_color.name(),
            "style": self.combo_style.currentText(),
            "animation": self.current_animation,
        }

File: cpp/src/gui/dialogs/export_dialog.cpp
/*
 * NC-KTV GUI — Export Dialog Implementation
 */

#include "export_dialog.h"

#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QFormLayout>
#include <QFileDialog>
#include <QMessageBox>

namespace ncktv {

ExportDialog::ExportDialog(QWidget* parent)
    : QDialog(parent)
{
    setWindowTitle("Export");
    setMinimumSize(540, 480);
    setModal(true);
    setupUi();
    applyTheme();
}

void ExportDialog::setupUi() {
    auto* mainLayout = new QVBoxLayout(this);
    mainLayout->setSpacing(14);
    mainLayout->setContentsMargins(24, 20, 24, 20);

    // ── Title ────────────────────────────────────────────────────────────
    auto* titleLabel = new QLabel("🎬  Export Project", this);
    titleLabel->setStyleSheet("font-size:18px; font-weight:bold; color:#e0e0e0; border:none;");
    mainLayout->addWidget(titleLabel);

    // ── Preset Group ─────────────────────────────────────────────────────
    auto* presetGroup = new QGroupBox("Export Preset", this);
    auto* presetLayout = new QFormLayout(presetGroup);
    presetLayout->setSpacing(10);
    presetLayout->setContentsMargins(16, 20, 16, 12);

    m_presetCombo = new QComboBox(this);
    m_presetCombo->addItems({
        "Karaoke Video (with video + lyrics)",
        "Lyrics Video (lyrics over black)",
        "Karaoke Audio Only (no video)",
        "Subtitles Only (ASS/SRT)"
    });
    presetLayout->addRow("Preset:", m_presetCombo);
    mainLayout->addWidget(presetGroup);

    // ── Format & Quality Group ───────────────────────────────────────────
    auto* formatGroup = new QGroupBox("Format && Quality", this);
    auto* formatForm = new QFormLayout(formatGroup);
    formatForm->setSpacing(10);
    formatForm->setContentsMargins(16, 20, 16, 12);

    m_formatCombo = new QComboBox(this);
    m_formatCombo->addItems({"MP4 (H.264)", "MKV (H.265)", "WebM (VP9)", "MP3 (Audio Only)", "ASS Subtitles", "SRT Subtitles"});
    formatForm->addRow("Output Format:", m_formatCombo);

    m_resolutionCombo = new QComboBox(this);
    m_resolutionCombo->addItems({"1920×1080 (Full HD)", "1280×720 (HD)", "3840×2160 (4K)", "Custom..."});
    formatForm->addRow("Resolution:", m_resolutionCombo);

    m_audioSourceCombo = new QComboBox(this);
    m_audioSourceCombo->addItems({"Instrumental", "Original (Mixed)", "Vocals Only"});
    formatForm->addRow("Audio Source:", m_audioSourceCombo);

    mainLayout->addWidget(formatGroup);

    // ── Lyrics Options Group ─────────────────────────────────────────────
    auto* lyricsGroup = new QGroupBox("Lyrics Options", this);
    auto* lyricsForm = new QFormLayout(lyricsGroup);
    lyricsForm->setSpacing(10);
    lyricsForm->setContentsMargins(16, 20, 16, 12);

    m_lyricsStyleCombo = new QComboBox(this);
    m_lyricsStyleCombo->addItems({
        "Neon Gold",
        "Classic Blue",
        "Modern Clean",
        "Fire Red",
        "Match Preview"
    });
    lyricsForm->addRow("Lyrics Style:", m_lyricsStyleCombo);

    m_burnSubsCheck = new QCheckBox("Burn subtitles into video (hardcoded)", this);
    m_burnSubsCheck->setChecked(true);
    lyricsForm->addRow("", m_burnSubsCheck);

    mainLayout->addWidget(lyricsGroup);

    // ── Output Path ──────────────────────────────────────────────────────
    auto* outputGroup = new QGroupBox("Output", this);
    auto* outputLayout = new QHBoxLayout(outputGroup);
    outputLayout->setContentsMargins(16, 20, 16, 12);

    m_outputEdit = new QLineEdit(this);
    m_outputEdit->setPlaceholderText("Select output location...");
    m_outputEdit->setReadOnly(true);
    m_browseBtn = new QPushButton("Browse...", this);
    m_browseBtn->setFixedWidth(90);
    outputLayout->addWidget(m_outputEdit, 1);
    outputLayout->addWidget(m_browseBtn);
    mainLayout->addWidget(outputGroup);

    // ── Buttons ──────────────────────────────────────────────────────────
    mainLayout->addStretch();

    auto* buttonLayout = new QHBoxLayout();
    buttonLayout->addStretch();

    m_cancelBtn = new QPushButton("Cancel", this);
    m_cancelBtn->setFixedWidth(100);
    buttonLayout->addWidget(m_cancelBtn);

    m_exportBtn = new QPushButton("Export", this);
    m_exportBtn->setFixedWidth(130);
    m_exportBtn->setDefault(true);
    m_exportBtn->setStyleSheet("background:#16a34a; color:white; font-weight:bold; border:none; padding:8px 16px; border-radius:4px;");
    buttonLayout->addWidget(m_exportBtn);

    mainLayout->addLayout(buttonLayout);

    // ── Connections ──────────────────────────────────────────────────────
    connect(m_browseBtn, &QPushButton::clicked, this, &ExportDialog::browseOutput);
    connect(m_exportBtn, &QPushButton::clicked, this, &ExportDialog::onAccepted);
    connect(m_cancelBtn, &QPushButton::clicked, this, &QDialog::reject);
    connect(m_presetCombo, QOverload<int>::of(&QComboBox::currentIndexChanged),
            this, &ExportDialog::onPresetChanged);
}

void ExportDialog::applyTheme() {
    setStyleSheet(R"(
        QDialog { background:#1e1e1e; color:#d4d4d4; }
        QGroupBox { border:1px solid #3f3f46; border-radius:6px; padding-top:16px; margin-top:8px; font-weight:bold; color:#ccc; }
        QGroupBox::title { subcontrol-origin:margin; left:12px; padding:0 6px; }
        QLineEdit { background:#2d2d30; border:1px solid #3f3f46; border-radius:4px; padding:6px 8px; color:#fff; }
        QComboBox { background:#2d2d30; border:1px solid #3f3f46; border-radius:4px; padding:6px 8px; color:#fff; }
        QComboBox::drop-down { border:none; }
        QPushButton { background:#333; border:1px solid #555; border-radius:4px; padding:6px 14px; color:#ccc; }
        QPushButton:hover { background:#404040; }
        QCheckBox { color:#ccc; }
        QLabel { color:#ccc; border:none; }
    )");
}

void ExportDialog::browseOutput() {
    QString fmt = m_formatCombo->currentText();
    QString ext = fmt.contains("ASS")  ? "ass"
                : fmt.contains("SRT")  ? "srt"
                : fmt.contains("MKV")  ? "mkv"
                : fmt.contains("WebM") ? "webm"
                : fmt.contains("MP3")  ? "mp3"
                : "mp4";

    QString filter = QString("Output File (*.%1);;All Files (*)").arg(ext);
    QString path = QFileDialog::getSaveFileName(this, "Save Output As", "", filter);
    if (!path.isEmpty()) {
        m_outputEdit->setText(path);
    }
}

void ExportDialog::onPresetChanged(int index) {
    // Adjust available options based on preset
    bool isVideoExport = (index <= 1);
    bool isSubOnly     = (index == 3);
    bool isAudioOnly   = (index == 2); // "Karaoke Audio Only"

    m_resolutionCombo->setEnabled(isVideoExport);
    m_burnSubsCheck->setEnabled(isVideoExport);
    m_audioSourceCombo->setEnabled(!isSubOnly);

    if (isSubOnly) {
        m_formatCombo->setCurrentText("ASS Subtitles");
    } else if (isAudioOnly) {
        m_formatCombo->setCurrentText("MP3 (Audio Only)");
    }
}

void ExportDialog::onAccepted() {
    if (m_outputEdit->text().isEmpty()) {
        QMessageBox::warning(this, "No Output", "Please select an output file path.");
        return;
    }
    accept();
}

QString ExportDialog::outputPath()   const { return m_outputEdit->text(); }
QString ExportDialog::format()       const { return m_formatCombo->currentText(); }
QString ExportDialog::preset()       const { return m_presetCombo->currentText(); }
QString ExportDialog::lyricsStyle()  const { return m_lyricsStyleCombo->currentText(); }
QString ExportDialog::audioSource()  const { return m_audioSourceCombo->currentText(); }

int ExportDialog::videoWidth() const {
    QString res = m_resolutionCombo->currentText();
    if (res.contains("3840")) return 3840;
    if (res.contains("1280")) return 1280;
    return 1920;
}

int ExportDialog::videoHeight() const {
    QString res = m_resolutionCombo->currentText();
    if (res.contains("2160")) return 2160;
    if (res.contains("720"))  return 720;
    return 1080;
}

bool ExportDialog::burnSubtitles() const { return m_burnSubsCheck->isChecked(); }

} // namespace ncktv
